File: unilingo-backend/app/schemas/vocabulary.py
"""
Pydantic schemas for Vocabulary
"""
import uuid
from datetime import datetime
from pydantic import BaseModel, Field


class AddVocabularyRequest(BaseModel):
    word: str = Field(..., min_length=1, max_length=100)
    phonetic: str | None = None
    audio_url: str | None = None
    definitions: list[dict] | None = None
    examples: list[str] | None = None
    user_note: str | None = None
    source_context: str | None = None
    source_attempt_id: uuid.UUID | None = None
    tags: list[str] | None = None


class UpdateVocabularyRequest(BaseModel):
    user_note: str | None = None
    examples: list[str] | None = None
    mastery_level: str | None = Field(None, pattern=r"^(new|learning|reviewing|mastered)$")
    tags: list[str] | None = None


class VocabularyResponse(BaseModel):
    id: uuid.UUID
    word: str
    phonetic: str | None
    audio_url: str | None
    definitions: list[dict] | None
    examples: list[str] | None
    user_note: str | None
    source_context: str | None
    mastery_level: str
    review_count: int
    next_review_at: datetime | None
    tags: list[str]
    created_at: datetime

    model_config = {"from_attributes": True}


class VocabularyListResponse(BaseModel):
    items: list[VocabularyResponse]
    total: int
    page: int
    per_page: int


class DictionaryLookupResponse(BaseModel):
    word: str
    phonetic: str | None
    audio_url: str | None
    meanings: list["DictionaryMeaning"]


class DictionaryMeaning(BaseModel):
    part_of_speech: str
    definitions: list["DictionaryDefinition"]


class DictionaryDefinition(BaseModel):
    definition: str
    example: str | None = None
    synonyms: list[str] = []
    antonyms: list[str] = []

File: unilingo_frontend/src/api/vocabulary.ts
/**
 * Vocabulary API service
 */
import apiClient from './client';

export interface VocabularyItem {
  id: string;
  word: string;
  phonetic: string | null;
  audio_url: string | null;
  definitions: any;
  examples: any;
  user_note: string | null;
  source_context: string | null;
  mastery_level: 'new' | 'learning' | 'reviewing' | 'mastered';
  review_count: number;
  next_review_at: string | null;
  tags: string[];
  created_at: string;
}

export interface AddVocabularyPayload {
  word: string;
  phonetic?: string;
  audio_url?: string;
  definitions?: any;
  examples?: any;
  user_note?: string;
  source_context?: string;
  source_attempt_id?: string;
  tags?: string[];
}

export interface DictionaryResult {
  word: string;
  phonetic: string | null;
  audio_url: string | null;
  meanings: {
    part_of_speech: string;
    definitions: {
      definition: string;
      example: string | null;
      synonyms: string[];
      antonyms: string[];
    }[];
  }[];
}

export const vocabularyAPI = {
  list: async (params?: {
    page?: number;
    per_page?: number;
    mastery_level?: string;
    search?: string;
    sort_by?: string;
    order?: string;
  }): Promise<{ items: VocabularyItem[]; total: number; page: number; per_page: number }> => {
    const { data } = await apiClient.get('/vocabulary', { params });
    return data;
  },

  add: async (payload: AddVocabularyPayload): Promise<VocabularyItem> => {
    const { data } = await apiClient.post('/vocabulary', payload);
    return data;
  },

  update: async (id: string, payload: Partial<AddVocabularyPayload>) => {
    const { data } = await apiClient.patch(`/vocabulary/${id}`, payload);
    return data;
  },

  delete: async (id: string) => {
    await apiClient.delete(`/vocabulary/${id}`);
  },

  getReviewDue: async () => {
    const { data } = await apiClient.get('/vocabulary/review-due');
    return data;
  },

  lookupDictionary: async (word: string): Promise<DictionaryResult> => {
    const { data } = await apiClient.get('/vocabulary/dictionary/lookup', {
      params: { word },
    });
    return data;
  },
};
